### DISCUSSION TITLE
ArrayBuffer concurrent write safety question

### DISCUSSION THREAD

**Discussion by Sytten** [nodejs/help#4621] [Q&A] [answered]
Hi!
Newbie question here (Rust background), I am reading the code for `fs` and I am wondering how is it safe for node to take the `ArrayBuffer` data and pass down the pointer to libuv async? Is it expected that there will be garbage in the buffer if you have two concurrent `Read` calls using the same `ArrayBuffer`? I am likely missing something here.
https://github.com/nodejs/node/blob/059f2f43b814619cdc0f93cf3683bd8f5c349c6a/src/node_file.cc#L2475-L2476
https://github.com/nodejs/node/blob/059f2f43b814619cdc0f93cf3683bd8f5c349c6a/src/node_file.cc#L2497-L2501
Thanks!

**Comment by avivkeller** [comment]
@nodejs/fs @nodejs/libuv

**Comment by mateuspaulino** [comment, marked as answer]
Hi @Sytten 

Great question! When you see the Node.js code like:


```
char* buffer_data = Buffer::Data(buffer_obj);
size_t buffer_length = Buffer::Length(buffer_obj);
```

and then later:


```
FSReqBase* req_wrap_async = GetReqWrap(args, 5);
CHECK_NOT_NULL(req_wrap_async);
FS_ASYNC_TRACE_BEGIN0(UV_FS_READ, req_wrap_async)
AsyncCall(env, req_wrap_async, args, "read", UTF8, AfterInteger, uv_fs_read, fd, &uvbuf, 1, pos);
```

you might wonder about the safety concerning concurrent reads using the same ArrayBuffer. Here's what's happening:
Memory Management and Safety

Reference Counting: Node.js uses a garbage-collected environment (V8 engine). When you pass a buffer to an async libuv operation, Node.js internally increments the reference count for the buffer. This ensures that the buffer will not be garbage collected while the asynchronous operation is pending. The buffer will only be freed when the reference count drops to zero, i.e., after the async operation completes and no other references to the buffer exist.

Concurrency: If you make two concurrent read calls using the same ArrayBuffer, both calls will indeed use the same memory space. This is generally unsafe because the reads can overwrite each other, leading to race conditions and corrupted data. Therefore, it is up to the developer to ensure that the same buffer is not used for concurrent read or write operations. Proper handling of concurrent operations (using locks, different buffers, or other synchronization mechanisms) is necessary to prevent such issues.

Internal Handling: The `FSReqBase` and associated async mechanisms in Node.js ensure that the async operation holds a reference to the buffer until the operation completes. The AsyncCall function sets up the necessary environment for the async operation, including proper reference counting and error handling.

Practical Implications

Avoid Concurrent Access: You should avoid using the same buffer for multiple concurrent read/write operations. Each async operation should ideally work on a separate buffer to prevent data corruption.

Libuv's Role: Libuv handles the actual asynchronous I/O operations. It takes the pointer to the buffer and performs the read/write in the background. Node.js ensures the buffer is valid during this operation by maintaining references.

Node.js and libuv are designed to work safely with buffers in asynchronous operations, but concurrent access to the same buffer must be managed by the developer. Ensuring separate buffers for each concurrent operation or implementing proper synchronization will help maintain data integrity.

Hope this clears up your doubts!

**Reply by Sytten** [reply]
Thank you very much for the detailed answer @mateuspaulino, it is appreciated!
Hopefully this can also help other people understand the source better.
For context I am rewriting some [APIs like FileHandle in rust for the QuickJS engine](https://github.com/awslabs/llrt/blob/3c45aa89a6c6600bf100141a3ada29c15a6135f9/llrt_modules/src/modules/fs/file_handle.rs) and reading node code to understand the subtleties of the API.
